### PR TITLE
Fix setup.py and use of requirements files and add a docker test container

### PR DESCRIPTION
**Pull request by jeking3**
Before:
```
jking@ubuntu:~/GitPython$ python3 setup.py check
/usr/lib/python3.6/distutils/dist.py:261: UserWarning: Unknown distribution option: 'test_requirements'
  warnings.warn(msg)
running check
```

Note that `setup.py` was not even using the contents of `requirements.txt` or `test-requirements.txt`.  Now it does, and it properly identifies the `tests-require` field.  The `requirements.txt` file declared a runtime dependency on a testing package.  I'm assuming that is an error (although it could be used in production code, I don't know...) so I removed that.

When I ran tox locally on Python 3.6.7 I saw a flake8 error F821 which I disabled as it looks bogus - that exception is defined in the documentation.

Added a docker container that you can use with a couple new make targets.  This allows you to run python 2.7, 3.4, 3.5, 3.6, or 3.7 tox even if you are on a local branch not named "master" (some tests require this).

Finally, found and fixed the issue on CI with python 3.7 / Xenial / travis.  The root cause is that in git 2.20 new behavior was introduced to stop "git fetch --tags" from clobbering local tags that have changed: https://github.com/git/git/commit/0bc8d71b99e91c9e90b519073b639a5066119591

The unit tests expect the old behavior, so now we pass --force down if the version is 2.20 or higher (for the remote test only, the code change is in the test code).  It took a while to find this because for some reason when GitCommandError returns an error code, stderr isn't captured.  If it was, it would have been easier to figure this out...

### CODE REVIEW

**Comment by jeking3** [comment]
It looks like CI isn't healthy.  Do you want some help with that?

**Comment by Byron** [comment]
Thanks a bunch for uncovering this, and for the fix!

It looks like one travis failure legit, as [it is flake not liking some whitespace](https://travis-ci.org/gitpython-developers/GitPython/jobs/480094618#L1032).
The [other travis failure](https://travis-ci.org/gitpython-developers/GitPython/jobs/480094621#L719) looks like it is related to the - now - flaky Remote test, which also fails consistently in #825 .

Please have a look at the comment there - maybe we can make a fix here so #825 can then be merged, too.

CC @cclauss

File: git/cmd.py [hunk 1]
```diff
@@ -220,7 +220,8 @@ def refresh(cls, path=None):
             # - a GitCommandNotFound error is spawned by ourselves
             # - a PermissionError is spawned if the git executable provided
             #   cannot be executed for whatever reason
-            exceptions = (GitCommandNotFound, PermissionError)
+            exceptions = (GitCommandNotFound, PermissionError)  # noqa
+            #                           (silence erroneous flake8 F821)
```
[the remark this answers is not in the document]
Comment (jeking3) [reply git/cmd.py]: Thanks, I did that.

**Comment by cclauss** [comment]
Run Python 3.7 in __allow_failures__ mode until #827 is fixed?

**Comment by jeking3** [comment]
What about the windows failures?

**Comment by cclauss** [comment]
```
FileNotFoundError: [WinError 2] The system cannot find the file specified: 'C:\\Miniconda35-x64\\lib\\site-packages\\setuptools-27.2.0-py3.5.egg'
```
But the lines just above that say that the [current version of setuptools (40.6.3)](https://pypi.org/project/setuptools/) was just installed.  Either don't upgrade setup tools or eliminate the hardcoding of  setuptools-27.2.0.

**Comment by jeking3** [comment]
What I was asking is do you want to ignore those too?

I'd prefer to merge this, and then deal with the CI failures, instead of bundling them all together... these changes improve (and fix) the setup.py issues.

**Comment by jeking3** [comment]
Do we need the last three windows builds for "miniconda" or "git running in cygwin on windows"?  The latter just looks like a test environment failure as it is passing windows paths (from the windows native python 3.5-x64) into a program looking for a unix path.  I have no idea what the "conda" ones are or if they provide any real value.

**Comment by jeking3** [comment]
The test suite makes a very bad assumption that you are on the master branch.  When I work on a project, I fork it, then I make a branch for each of the changes I am making, and submit pull requests from the branch into the upstream master.  This is best practices for github development.

I cannot run the test suite if I have anything else checked out.  The test suite fails.  It is making some bad assumptions.  When I run inside a docker container and skip past those, even with python 3.6, I'm seeing other issues like utf8 issues.

**Comment by jeking3** [comment]
I worked around the issues by checking out a copy of the repository from github into the container at a well-known location.  This way the unit tests can run against a fresh clone of the repository and succeed.

**Comment by jeking3** [comment]
I'm going to tease this apart into smaller bits...

**Comment by jeking3** [comment]
@Byron consider merging this.   The appveyor jobs are passing, at least the ones known to pass, and there are some that are known failures; it would still be nice to get all the CI builds to green since that makes it so much easier to determine if a pull request is even worth looking into.

**Comment by Byron** [comment]
Hi @jeking3 , thanks so much for your persistency. Right now I am very short on time and am far away from this, and any other of my open source projects, which must be frustrating for GitPython's user base.

Maybe it would help if there were more collaborators - I have invited you in case this sounds feasible to you.

**Comment by jeking3** [comment]
I'll do what I can.  Apart from founding a new company, I'm also involved in boost, apache thrift, and a couple other efforts.  At the very least if I can get the CI builds to run clean and have folks rebase their pull requests, it would be a lot easier to qualify things in the pipeline.  Thanks!

**Pull request merged**